server/src/main/java/org/opensearch/action/search/SearchResponse.java

```diff
@@ -593,4 +633,90 @@ static SearchResponse empty(Supplier<Long> tookInMillisSupplier, Clusters cluste
             null
         );
     }
+
+    /**
+     * SearchResponse builder that provides an easy way to make a copy of a search response
+     * while replacing only certain (internal) parts of it.
+     *
+     * SearchResponse newResponse = new SearchResponse.Builder().from(oldResponse).hits(newHits).build();
+     *
+     */
+    public static class Builder {
```

Comment:
I see this class is only used in tests, I don't see why it should be part of the public API, please move it there

Reply:
The idea here is for the SearchResponse class to own the logic of constructing itself using its various internals.  Today, if you want to create a response from another response because you need to replace one thing, you have to manually copy over each individual piece.

Take a look at the SearchResponseUtil here - https://github.com/opensearch-project/OpenSearch/pull/9405/files.

I will also have to do something similar in ml-commons where a search processor will be adding SearchExtBuilder to SearchResponse.  I didn't want a bunch of util classes doing the same thing in multiple places.  SearchResponse seems like the right place.  You just need to update the Builder every time you are introducing something new to SearchResponse.

Reply:
I feel very uncomfortable about that: this class does not copy but share and, sadly, the search response are section are mutable. We also have discussion opened here https://github.com/opensearch-project/OpenSearch/issues/869 if you want to bring this one up, but I would suggest to remove from the scope of this change.

Reply:
Somehow, I commented on that issue a while ago as well.  :-)

This change is more about not exposing the internals of SearchResponse so that you can make a copy without having to know what to copy over.  This does not introduce any new (deep) copy logic.

I will remove it.  We can always bring it back if we need it later.  Thanks.